### PR TITLE
Small fixes for emoticons like :-)

### PR DESCRIPTION
**Pull request by jexe**
Hi,

Put in a few small fixes for emoticon support. Let me know what you think? Thx!

### CODE REVIEW

**Comment by valeriomazzeo** [comment]
Hi,

Thanks for the pull request.

What's the benefit of having the emoticon map in a separate array?

**Comment by jexe** [comment]
It's basically a bug fix: when doing string matching, we weren't able to resolve :octopus: because the :o emoticon was being matched and transformed first. Depending on how the map gets built I could imagine others like :p having similar issues down the road as the list grows. 

Also- because of a performance optimization, semicolon-based faces like ;-) weren't working. That's been fixed here as well.

**Comment by valeriomazzeo** [comment]
Interesting, I will have a look soon just to double check and then merge :)

Thanks again!

**Comment by jexe** [comment]
Cool. Thanks for considering it!

File: NSString+Emoji.m [hunk 1]
```diff
@@ -10,19 +10,34 @@
...
+                                  @":-)": @"😃",
+                                  @":(": @"😞",
+                                  @":-(": @"😞",
+                                  @";)": @"😉",
+                                  @";-)": @"😉",
+                                  @":P": @"😜",
+                                  @":-P": @"😜",
+                                  @":o": @"😮",
+                                  @":-o": @"😮",
+                                  };
```
Comment (valeriomazzeo) [inline NSString+Emoji.m:32]: Can you please add a new line here?

**Comment by valeriomazzeo** [comment]
With your changes, when building the reverse map:

```
NSMutableDictionary *reversedMap = [NSMutableDictionary dictionaryWithCapacity:[forwardMap count]];
    [forwardMap enumerateKeysAndObjectsUsingBlock:^(id key, id obj, BOOL *stop) {
        if ([obj isKindOfClass:[NSArray class]]) {
            for (NSString *object in obj) {
                [reversedMap setObject:key forKey:object];
            }
        } else {
            [reversedMap setObject:key forKey:obj];
        }
    }];
```

We should not check for the existence of an array anymore, because you mapped all the emoticons into a different one so the mapping in this new case is only 1:1.

Also, the `emoticonMap` array has to be merged (after swapping keys/objects) into the reversed array unless the conversion in `stringByReplacingEmojiCheatCodesWithUnicode` won't work properly for the emoticons contained in the `emoticonMap` array.

Last thing, maybe we should consider creating the character set: `[NSCharacterSet characterSetWithCharactersInString:@":;"]`

dynamically into: `initializeEmojiCheatCodes`

So if in the future we add new emojis that start with a character different than `:`, `;`, it can be picked up correctly.

Let me know your thoughts!

**Comment by valeriomazzeo** [comment]
Hi @jexe, just a poke to understand if you are still interested in this.

**Comment by andlister** [comment]
Hey

Sorry for the delay. Thanks for the feedback.
1. I think we still need the array checking because emoji like :+1:, :-1: and :hankey: have more than one cheated code to unicode mappings.
2. I am not sure I am following you with merging the emoticonMap into the reversed array? The stringByReplacingEmojiCheatCodesWithUnicode is processing the s_emoticonToUnicode. 

Thanks

**Pull request closed without merging**